radicle/src/cob/identity.rs

```diff
@@ -0,0 +1,746 @@
+            .collect()
+    }
+
+    pub fn reaches_quorum(&self, previous: &Identity<Oid>) -> bool {
+        let votes_for = self
+            .verdicts
+            .iter()
+            .fold(0, |count, (_, verdict)| match verdict.get() {
+                Some(Verdict::Accept(_)) => count + 1,
+                Some(Verdict::Reject) => count,
+                None => count,
+            });
+        votes_for >= previous.doc.threshold
+    }
+
+    fn accept(&mut self, key: PublicKey, signature: Signature, clock: Clock) {
+        self.verdicts
+            .insert(key, Redactable::Present(Verdict::Accept(signature)), clock);
+    }
+
+    fn reject(&mut self, key: PublicKey, clock: Clock) {
+        self.verdicts
+            .insert(key, Redactable::Present(Verdict::Reject), clock)
+    }
+}
+
+impl store::Transaction<Proposal> {
+    pub fn accept(&mut self, revision: RevisionId, signature: Signature) -> OpId {
+        self.push(Action::Accept {
+            revision,
+            signature,
+        })
+    }
+
+    pub fn reject(&mut self, revision: RevisionId) -> OpId {
+        self.push(Action::Reject { revision })
+    }
+
+    pub fn edit(&mut self, title: impl ToString, description: impl ToString) -> OpId {
+        self.push(Action::Edit {
+            title: title.to_string(),
+            description: description.to_string(),
+        })
+    }
+
+    pub fn close(&mut self) -> OpId {
+        self.push(Action::Close)
+    }
+
+    pub fn publish(&mut self) -> OpId {
+        self.push(Action::Publish)
+    }
+
+    pub fn redact(&mut self, revision: RevisionId) -> OpId {
+        self.push(Action::Redact { revision })
+    }
+
+    pub fn revision(&mut self, current: Oid, proposed: Doc<Verified>) -> OpId {
+        self.push(Action::Revision { current, proposed })
+    }
+
+    /// Start a patch revision discussion.
+    pub fn thread<S: ToString>(&mut self, revision: RevisionId, body: S) -> OpId {
+        self.push(Action::Thread {
+            revision,
+            action: thread::Action::Comment {
+                body: body.to_string(),
+                reply_to: None,
+            },
+        })
+    }
+
+    /// Comment on a proposal revision.
+    pub fn comment<S: ToString>(
+        &mut self,
+        revision: RevisionId,
+        body: S,
+        reply_to: thread::CommentId,
+    ) -> OpId {
+        self.push(Action::Thread {
+            revision,
+            action: thread::Action::Comment {
+                body: body.to_string(),
+                reply_to: Some(reply_to),
+            },
+        })
+    }
+
+    /// Update a proposal with a new revision.
+    pub fn update(
+        &mut self,
+        title: impl ToString,
+        description: impl ToString,
+        current: impl Into<Oid>,
+        proposed: Doc<Verified>,
+    ) -> (OpId, OpId) {
+        let revision = self.revision(current.into(), proposed);
+        let meta = self.edit(title, description);
+
+        (revision, meta)
```

Comment:
I'm guessing this is the culprit of your revision problem.

Due to the `edit`, the clock is being incremented by one *after* the revision is created.
I'd consider splitting the edit outside of this function, or tying title/description to the revision.

Reply:
I think in the Patch cob, we have a seperate edit and update function, which avoids this problem.

Reply:
Ah, right. The `meta` is on the `Proposal` and not the revision itself! :facepalm: Couldn't I just return `revision`? The `OpId` for `meta` is kind of useless, or is it?

Reply:
Ended up splitting them out, makes sense now :)